gtsam/hybrid/IncrementalHybrid.cpp

```diff
@@ -88,133 +48,78 @@ void IncrementalHybrid::update(GaussianHybridFactorGraph graph,
-  for (size_t i = 0; i < bayesNetFragment->size(); i++) {
-    auto conditional = bayesNetFragment->at(i);
-
-    // Container for nodes (or nullptrs) to create a new DecisionTree
-    std::vector<GaussianConditional::shared_ptr> nodes;
-
-    // Loop over all assignments and create a set of GaussianConditionals
-    std::function<GaussianFactor::shared_ptr(
-        const Assignment<Key> &, const GaussianFactor::shared_ptr &)>
-        pruner = [&](const Assignment<Key> &choices,
-                     const GaussianFactor::shared_ptr &gf) {
-          // typecast so we can use this to get probability value
-          DiscreteValues values(choices);
-
-          if ((*discreteFactor)(values) == 0.0) {
-            // empty aka null pointer
-            boost::shared_ptr<GaussianFactor> null;
-            return null;
-          } else {
-            return gf;
+          } else if (auto df = boost::dynamic_pointer_cast<DiscreteConditional>(
+                         conditional)) {
+            graph.push_back(df);
+            conditionals_to_erase.push_back(conditional);
           }
-        };
-
-    GaussianMixture::shared_ptr gaussianMixture =
-        boost::dynamic_pointer_cast<GaussianMixture>(conditional);
-
-    if (gaussianMixture) {
-      // We may have mixtures with less discrete keys than discreteFactor so we
-      // skip those since the label assignment does not exist.
-      std::set<DiscreteKey> gmKeySet = gaussianMixture->discreteKeys().asSet();
-      std::set<DiscreteKey> dfKeySet = discreteFactor->discreteKeys().asSet();
-      if (gmKeySet != dfKeySet) {
-        continue;
-      }
-
-      // Run the pruning to get a new, pruned tree
-      auto prunedFactors = gaussianMixture->factors_.apply(pruner);
-
-      DiscreteKeys discreteKeys = gaussianMixture->discreteKeys();
-      // reverse keys to get a natural ordering
-      std::reverse(discreteKeys.begin(), discreteKeys.end());
-
-      // Convert to GaussianConditionals
-      auto prunedTree = GaussianMixture::Conditionals(
-          prunedFactors, [](const GaussianFactor::shared_ptr &factor) {
-            return boost::dynamic_pointer_cast<GaussianConditional>(factor);
-          });
 
-      // Create the new gaussian mixture and add it to the bayes net.
-      auto prunedGaussianMixture = boost::make_shared<GaussianMixture>(
-          gaussianMixture->nrFrontals(), gaussianMixture->continuousKeys(),
-          discreteKeys, prunedTree);
-
-      prunedBayesNetFragment->setGaussian(i, prunedGaussianMixture);
+          break;
+        }
+      }
+    }
+    // Remove conditionals at the end so we don't affect the order in the
+    // original bayes net.
+    for (auto &&conditional : conditionals_to_erase) {
+      auto it = find(hybridBayesNet.begin(), hybridBayesNet.end(), conditional);
+      hybridBayesNet.erase(it);
     }
   }
-
-  return prunedBayesNetFragment;
+  return std::make_pair(graph, hybridBayesNet);
```

Comment:
Nit {}

Reply:
Will update this in the `fix/pruning` PR since I want to add another change there as well. :slightly_smiling_face: